Added TestSetupMixin for testbed

svpino opened the pull request:
This PR contains the TestSetupMixin for the testbed. It also offers a teardown method to deactivate the testbed.

aozarov commented:
Did you consider this pattern:

class U(object):

  def setUp(self):
    print "U.setUp"

  def tearDown(self):
    print "U.tearDown"

class M(object):

  QUEUE = []

  def setUp(self):
    super(M, self).setUp()
    print "M.setUp " + str(self.QUEUE)

  def tearDown(self):
    super(M, self).tearDown()
    print "M.tearDown"


class C(M, U):

  QUEUE = ['other']

  def setUp(self):
    super(C, self).setUp()


c = C()
c.setUp()
c.tearDown()

svpino commented:
Hey @aozarov,

Thanks for the recommendation! I don't have a whole lot of experience with Python so of course _mixins_ are a foreign concept coming with a Java background. Your example makes total sense, and I refactored the code to use it just like that.

(When I saw that your class `M` is calling `super.setUp` although it inherits from `object` I thought it was a mistake... or I was drunk... True story.)

So take a look to the changes and let me know how is it looking now.

Thanks

aozarov commented on python/test/testutil.py:
```diff
@@ -25,56 +25,58 @@
 
-TEST_APP_ID = 'my-app-id'
-TEST_VERSION_ID = 'my-version.1234'
+  test_app_id = 'my-app-id'
```
This should stay uppercase

aozarov commented on python/test/pipeline_test.py:
```diff
@@ -54,14 +54,16 @@
     super(TestBase, self).setUp()
     self.maxDiff = 10**10
 
+  def tearDown(self):
```
No need to redefine/override if you just call super

aozarov commented on python/test/pipeline_test.py:
```diff
@@ -805,50 +807,46 @@ def testAbortRootAsyncAlreadyAborted(self):
 
-      result = []
-      def fake_mail(self, sender, subject, body, html=None):
+    self.test_app_id = 's~my-hrd-app'
```
I am not sure why this test really needed to change the app id (maybe it does not and instead validation should be done against the regular app id defined in testutil.
If you think it is needed you can seperate the method into its own test class and override the TEST_APP_ID constant (as I showed in the example).

aozarov commented on python/test/pipeline_test.py:
```diff
@@ -805,50 +807,46 @@ def testAbortRootAsyncAlreadyAborted(self):
-      result = []
-      def fake_mail(self, sender, subject, body, html=None):
+    self.test_app_id = 's~my-hrd-app'
+    super(PipelineTest, self).setUp()
```
should not be necessary.

aozarov commented on python/test/testutil.py:
```diff
@@ -25,56 +25,58 @@
-os.environ['HTTP_HOST'] = '%s.appspot.com' % TEST_APP_ID
-os.environ['DEFAULT_VERSION_HOSTNAME'] = os.environ['HTTP_HOST']
-os.environ['CURRENT_MODULE_ID'] = 'foo-module'
+  os.environ['APPLICATION_ID'] = test_app_id
```
Maybe this should move to the setUp as well (if we are really changing TEST_APP_ID).

aozarov commented:
Looks good. few minor comments

svpino commented:
Thanks @aozarov!

Here is what comes in the new commit:

1) TEST_APP_ID and TEST_VERSION_ID are back in uppercase.
2) Removed the unnecessary call to `super...` in `tearDown`
3) `testFinalizeEmailDone_HighReplication` does need to change the APP_ID. I extracted it to a separate class and it's now using the pattern from your example.
4) I moved the os.environ initialization to inside the `setUp` method.

Let me know your thoughts.

Thanks

aozarov commented:
Looks good. Thank you for doing that.
